cnd/src/network/comit_ln/swaps.rs

```diff
@@ -0,0 +1,602 @@
+            Some(local_swap_id) => local_swap_id,
+            None => return None,
+        };
+
+        let create_params = match self.swaps.get(&local_swap_id) {
+            Some(create_params) => create_params,
+            None => return None,
+        };
+
+        let shared_swap_id = SharedSwapId::default();
+        self.swap_ids.insert(local_swap_id, shared_swap_id.clone());
+
+        Some((shared_swap_id, create_params.clone()))
+    }
+
+    /// Bob moves a swap that was announced and pending creation to communicate
+    /// after receiving an announcement from Alice
+    pub fn move_pending_creation_to_communicate(
+        &mut self,
+        digest: &SwapDigest,
+        local_swap_id: LocalSwapId,
+        create_swap_params: HanEtherereumHalightBitcoinCreateSwapParams,
+    ) -> Result<(SharedSwapId, PeerId, T), Error> {
+        if self
+            .swaps
+            .insert(local_swap_id, create_swap_params)
+            .is_some()
+        {
+            return Err(Error::AlreadyExists);
+        }
+
+        let (peer, io) = match self.pending_creation.remove(&digest) {
+            Some(value) => value,
+            None => return Err(Error::WasNotPending),
+        };
+
+        let shared_swap_id = SharedSwapId::default();
+        self.swap_ids.insert(local_swap_id, shared_swap_id.clone());
+
+        Ok((shared_swap_id, peer, io))
+    }
+
+    /// Either role finalizes a swap that was in the communication phase
+    /// This also proceeds with clean up from the various _pending_ stores.
+    pub fn finalize_swap(
+        &mut self,
+        shared_swap_id: &SharedSwapId,
+    ) -> Result<(LocalSwapId, HanEtherereumHalightBitcoinCreateSwapParams), Error> {
+        let local_swap_id = match self.swap_ids.iter().find_map(|(key, value)| {
+            if *value == *shared_swap_id {
+                Some(key)
+            } else {
+                None
+            }
+        }) {
+            Some(local_swap_id) => local_swap_id,
+            None => return Err(Error::UnknownId),
+        };
+
+        let create_params = match self.swaps.get(&local_swap_id) {
+            Some(create_params) => create_params,
+            None => return Err(Error::NotFound),
+        };
+
+        self.pending_announcement
+            .retain(|_, id| *id != *local_swap_id);
+
+        let finalized_digest = create_params.clone().digest();
+
+        self.timestamps
+            .retain(|digest, _| *digest != finalized_digest);
+
+        Ok((*local_swap_id, create_params.clone()))
+    }
+
+    /// Remove all pending (not finalized) swap older than `older_than`
+    pub fn clean_up_pending_swaps(&mut self, older_than: Timestamp) {
+        let digests: Vec<SwapDigest> = self
+            .timestamps
+            .iter()
+            .filter(|(_, timestamp)| **timestamp < older_than)
+            .map(|(digest, _)| digest)
+            .cloned()
+            .collect();
```

Comment:
I always try to avoid type hints because they could change if you refactor the function. Giving type hints on the fn's that actually need them doesn't suffer from that :)

```suggestion
        let digests = self
            .timestamps
            .iter()
            .filter(|(_, timestamp)| **timestamp < older_than)
            .map(|(digest, _)| digest)
            .cloned()
            .collect::<Vec<_>>();
```

Reply:
You could also avoid the allocation (i.e. using `collect`) by using the iterator with a for loop :)

```rust
for digest in digest_iter {
	self.pending_confirmation.remove(digest);
	// ...
}
```